lxd/db/warnings.go

```diff
@@ -0,0 +1,395 @@
+	stmt, err := c.tx.Prepare(str)
+	if err != nil {
+		return err
+	}
+	defer stmt.Close()
+
+	_, err = stmt.Exec(status, UUID)
+	if err != nil {
+		return errors.Wrap(err, "Failed to update warning status")
+	}
+
+	return nil
+}
+
+// UpdateWarningMessage updates the warning with the given ID.
+func (c *ClusterTx) UpdateWarningMessage(UUID string, message string) error {
+	str := fmt.Sprintf("UPDATE warnings SET last_message=?, last_seen_date=?, count=count+1 WHERE uuid=?")
+	stmt, err := c.tx.Prepare(str)
+	if err != nil {
+		return err
+	}
+	defer stmt.Close()
+
+	_, err = stmt.Exec(message, time.Now(), UUID)
+	if err != nil {
+		return errors.Wrap(err, "Failed to update warning")
+	}
+
+	return nil
+}
+
+func (c *ClusterTx) doGetWarnings(q string, args ...interface{}) ([]Warning, error) {
+	stmt, err := c.tx.Prepare(q)
+	if err != nil {
+		return nil, err
+	}
+	defer stmt.Close()
+
+	// Result slice.
+	objects := make([]Warning, 0)
+
+	// Dest function for scanning a row.
+	dest := func(i int) []interface{} {
+		objects = append(objects, Warning{})
+		return []interface{}{
+			&objects[i].ID,
+			&objects[i].Node,
+			&objects[i].Project,
+			&objects[i].EntityTypeCode,
+			&objects[i].EntityID,
+			&objects[i].UUID,
+			&objects[i].TypeCode,
+			&objects[i].Status,
+			&objects[i].FirstSeenDate,
+			&objects[i].LastSeenDate,
+			&objects[i].LastMessage,
+			&objects[i].Count,
+		}
+	}
+
+	// Select.
+	err = query.SelectObjects(stmt, dest, args...)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to fetch warnings")
+	}
+
+	return objects, nil
+}
+
+// GetWarningsByProject returns all available warnings in the given project.
+func (c *ClusterTx) GetWarningsByProject(projectName string) ([]Warning, error) {
+	q := `SELECT warnings.id, IFNULL(nodes.name, "") AS node, IFNULL(projects.name, "") AS project, IFNULL(warnings.entity_type_code, -1), IFNULL(warnings.entity_id, -1), warnings.uuid, warnings.type_code, warnings.status, warnings.first_seen_date, warnings.last_seen_date, warnings.last_message, warnings.count
+	FROM warnings LEFT JOIN nodes ON warnings.node_id = nodes.id LEFT JOIN projects ON warnings.project_id = projects.id WHERE project=? ORDER BY warnings.last_seen_date`
+
+	return c.doGetWarnings(q, projectName)
+}
+
+// GetWarnings returns all available warnings.
+func (c *ClusterTx) GetWarnings() ([]Warning, error) {
+	q := `SELECT warnings.id, IFNULL(nodes.name, "") AS node, IFNULL(projects.name, "") AS project, IFNULL(warnings.entity_type_code, -1), IFNULL(warnings.entity_id, -1), warnings.uuid, warnings.type_code, warnings.status, warnings.first_seen_date, warnings.last_seen_date, warnings.last_message, warnings.count
+	FROM warnings LEFT JOIN nodes ON warnings.node_id = nodes.id LEFT JOIN projects ON warnings.project_id = projects.id ORDER BY warnings.last_seen_date`
+
+	return c.doGetWarnings(q)
+}
+
+// CreateWarning adds a new warning to the database.
+func (c *ClusterTx) CreateWarning(object Warning) (int64, error) {
```

Comment:
@monstermunchkin it seems like this function is redundant given we have `CreateOrUpdateWarning` above?

Also, could `CreateOrUpdateWarning` accept the `Warning` struct type argument, rather than individual arguments for clarity?

Reply:
> @monstermunchkin it seems like this function is redundant given we have CreateOrUpdateWarning above?

I'd prefer keeping those functions separate due to their length. Merging them would result in a rather long function IMHO.

> Also, could CreateOrUpdateWarning accept the Warning struct type argument, rather than individual arguments for clarity?

That was intentional because a lot of the values in the struct (e.g. ID, UUID, Count, first seen, last seen) would simply be ignored when creating the warning. I thought that approach was clearer.